pkg/controller/mutators/core/reconciler.go

```diff
@@ -159,68 +108,95 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 		return reconcile.Result{}, err
 	}
 
+	ingestionStatus := ctrlmutators.MutatorStatusError
+	// Encasing this call in a function prevents the arguments from being evaluated early.
+	id := types.MakeID(mutationObj)
+	defer func() {
+		r.reportMutator(id, ingestionStatus, startTime)
+	}()
+
+	// previousConflicts records the conflicts this Mutator has with other mutators
+	// before making any changes.
+	previousConflicts := r.system.GetConflicts(id)
+	delete(previousConflicts, id)
+
 	if deleted {
 		// Either the mutator was deleted before we were able to process this request, or it has been marked for
 		// deletion.
-		return r.reconcileDeleted(ctx, mutationObj)
+		r.getTracker().CancelExpect(mutationObj)
+		err = r.reconcileDeleted(ctx, id)
+	} else {
+		err = r.reconcileUpsert(ctx, id, mutationObj)
 	}
 
-	ingestionStatus := ctrlmutators.MutatorStatusError
-	// encasing this call in a function prevents the arguments from being evaluated early
-
-	defer func() {
-		r.reportMutator(types.MakeID(mutationObj), ingestionStatus, startTime)
-	}()
-
-	mutator, err := r.mutatorFor(mutationObj)
 	if err != nil {
-		r.log.Error(err, "Creating mutator for resource failed", "resource", request.NamespacedName)
-		r.getTracker().TryCancelExpect(mutationObj)
-
-		err = r.updateStatus(ctx, mutationObj, false, err)
 		return reconcile.Result{}, err
 	}
 
-	if err = r.system.Upsert(mutator); err != nil {
-		r.log.Error(err, "Insert failed", "resource", request.NamespacedName)
-		r.getTracker().TryCancelExpect(mutationObj)
+	newConflicts := r.system.GetConflicts(id)
+	delete(newConflicts, id)
 
-		err = r.updateStatus(ctx, mutationObj, false, err)
-		return reconcile.Result{}, err
-	}
+	// diff is the set of mutators which either:
+	// 1) previously conflicted with mutationObj but do not after this change, or
+	// 2) now conflict with mutationObj but did not before this change.
+	diff := symmetricDifference(previousConflicts, newConflicts)
 
-	r.getTracker().Observe(mutationObj)
+	// Now that we've made changes to the recorded Mutator schemas, we can re-check
```

Comment:
I think a different name might better portray the implementation decision made here.  My understanding of what's going on is that we are finding the set of Mutators that now are in conflict, then trigger reconciliations for those mutators.  To minimize unnecessary reconciliations, we remove both `previousConflicts` (as they'd have already been reconciled) and the current `ID`, as we've just reconciled it.

`queueConflicts()` might be better named `triggerReconciliations()` or something like that.

It might also be nice to explain in a comment why we didn't take the approach of updating all the statuses.  This yielded some complexity in the implementation, and documenting why that decision was made could be very helpful to a future engineer.

Reply:
A lot to unpack here. One comment on the comment:

> It might also be nice to explain in a comment why we didn't take the approach of updating all the statuses. This yielded some complexity in the implementation, and documenting why that decision was made could be very helpful to a future engineer.

This sounds like the start of a best practices for authoring controllers doc. I wonder if one already exists?

Reply:
The comment for "queueConflicts" answers this:
```
// queueConflicts queues updates for Mutators in ids.
// We send events to the handler's event queue rather than attempting the update
// ourselves to delegate handling failures to the existing controller logic.
```